packages/plugin-core/src/commands/MoveHeader.ts

```diff
@@ -0,0 +1,311 @@
+  DendronError,
+  NoteProps,
+  NoteUtils,
+  VaultUtils,
+  getSlugger,
+} from "@dendronhq/common-all";
+import {
+  HistoryService,
+  DendronASTDest,
+  DendronASTTypes,
+  AnchorUtils,
+  Processor,
+  RemarkUtils,
+  MDUtilsV5,
+  ProcMode,
+  visit,
+  Heading,
+  Node,
+} from "@dendronhq/engine-server";
+import { LookupControllerV3 } from "../components/lookup/LookupControllerV3";
+import {
+  NoteLookupProvider,
+  NoteLookupProviderSuccessResp,
+} from "../components/lookup/LookupProviderV3";
+import { DENDRON_COMMANDS } from "../constants";
+import { VSCodeUtils } from "../utils";
+import { BasicCommand } from "./base";
+import _ from "lodash";
+import path from "path";
+import { getEngine } from "../workspace";
+import { EngineAPIService } from "../services/EngineAPIService";
+import { delayedUpdateDecorations } from "../features/windowDecorations";
+import { findReferences } from "../utils/md";
+
+type CommandInput =
+  | {
+      dest?: NoteProps;
+    }
+  | undefined;
+type CommandOpts = {
+  origin: NoteProps;
+  nodesToMove: Node[];
+  modifiedOriginTree: Node;
+  originProc: Processor;
+  engine: EngineAPIService;
+} & CommandInput;
+type CommandOutput = {
+  updated: NoteProps[];
+} & CommandOpts;
+
+export class MoveHeaderCommand extends BasicCommand<
+  CommandOpts,
+  CommandOutput
+> {
+  key = DENDRON_COMMANDS.MOVE_HEADER.key;
+
+  headerNotSelectedError = new DendronError({
+    message: "You must first select the header you want to move.",
+  });
+
+  noActiveNoteError = new DendronError({
+    message: "No note open.",
+  });
+
+  getProc = (engine: EngineAPIService, note: NoteProps) => {
+    return MDUtilsV5.procRemarkParse(
+      { mode: ProcMode.FULL },
+      {
+        engine,
+        fname: note.fname,
+        vault: note.vault,
+        dest: DendronASTDest.MD_DENDRON,
+      }
+    );
+  };
+
+  /**
+   * Recursively check if two given node is identical.
+   * At each level _position_ is omitted as this can change if
+   * you are comparing from two different trees.
+   * @param a first {@link Node} to compare
+   * @param b second {@link Node} to compare
+   * @returns boolean
+   */
+  hasIdenticalChildren = (a: Node, b: Node): boolean => {
+    if (_.isEqual(Object.keys(a).sort(), Object.keys(b).sort())) {
+      const aOmit = _.omit(a, ["position", "children"]);
+      const bOmit = _.omit(b, ["position", "children"]);
+      if (_.isEqual(aOmit, bOmit)) {
+        if (_.has(a, "children")) {
+          return _.every(a.children as Node[], (aChild, aIndex) => {
+            const bChild = (b.children as Node[])[aIndex];
+            return this.hasIdenticalChildren(aChild, bChild);
+          });
+        }
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  };
+
+  async gatherInputs(opts: CommandInput): Promise<CommandOpts | undefined> {
+    const engine = getEngine();
+    const { editor, selection } = VSCodeUtils.getSelection();
```

Comment:
dumb question, I forgot - does at least one character need to be selected for this to work? will this return anything if I just put my cursor in the header but nothing is selected?

Reply:
I believe so. we have the problem right now we're putting the cursor on a line for `CopyNoteLink` grabs the header (often when I don't mean to do it)